### PR TITLE
StreamData blog post

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+```elixir
+property "length/1 is always >= 0" do
+  check all list <- list_of(term()) do
+    assert length(list) >= 0
+  end
+end
+```
+
+With property-based testing, you specify a set of valid inputs (that is, lists in the example above) for your code and verify that your code holds some property for values taken at random from the valid inputs. In the example above, the test takes many (usually around 100) values at random from the `list_of(term())` *generator* and verifies that the property of `length/1` always returning a non-negative integer holds. A generator is just a `StreamData` generator, as we discussed in the previous section.
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: (~that is,~ lists in the example above)

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+### Shrinking
+
+Since we're generating many random data, if we hit a failure then the data that caused it is often complex and convolute. Take this trivial example of a property:
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Since we're generating lots of random inputs to test, the input that caused the failure is often complex and convoluted. Take this trivial example of a property:

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    end
+  end
+end
+```
+
+There's not much more to the mechanics of `StreamData`. Most of the work you will have to do revolves around finding good properties to test for your code and writing good generators for the data over which you want to test. Head over to [stream_data][]'s documentation for detailed documentation.
+
+### Advantages of property-based testing
+
+Using property-based testing has some advantages. First of all, it lets you test properties of your code over many more values that you otherwise would with example-based testing. While it's true that random data generation can't cover all the possible values that a piece of code can deal with, the confidence in your codebase can still increase over time because the property-based tests will likely generate different values on each run. Example-based testing means your test data will not change over time.
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: your code over many more values ~that~than you otherwise would

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+## Why include property-based testing in Elixir (and rewriting from scratch)
+
+The community has expressed some concern regarding two main things: why do we want to include a property-based testing tool in Elixir's standard library? And why write such a tool from scratch instead of using one of the existing Erlang or Elixir solutions?
+
+The answer to the first question is that we want to include such a tool in Elixir because we believe that providing such a tool in the standard library will encourage developers to use property-based testing and ultimately improve their software and the way they write it. At the same time, we want to be able to use property-based testing to test the Elixir codebase itself (which already turned out great [in the past][quickcheck-pr]). To do this, we can't rely on an external library, so we need to have a solution built into the standard library.
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: The answer to the first question is that we ~want to include such a tool in Elixir because we~ believe ~that~ providing such a tool in the standard library will encourage developers to use property-based testing and ultimately improve their software and the way they write it.

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+The reasons for writing such a tool from scratch are best explained by José in [this ElixirForum post][elixirforum-post]:
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: ~such a tool~ a new property-based testing library

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+In this blog post, we'll cover what is property-based testing and how it can benefit your programs, why we want it in Elixir, and what we are exploring for the future of it. If you want to use the features discussed below or you want to read more formal documentation, head over to [stream_data][].
+
+## Sample data generation
+
+The core of the [stream_data][] library is `StreamData`: this module (which will be renamed to `Stream.Data` once the library is merged into Elixir) provides all the functionalities related to generating sample data of many kinds. It includes both data generators for data types (like integers or booleans) as well as tools to combine other generators (such as `one_of(list_of_generators)`).
+
+Developers are not supposed to create generators from scratch, but use the provided generators and the provided ways to combine them. An example of a generator is the one returned by `StreamData.integer()`: this function returns a generator that generates integers. Generators implement the `Enumerable` protocol and they are infinite streams of terms. This means we can do operations such as taking terms out of a generator through functions from `Enum` and `Stream`:
```
Comment (ericmj) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: ~the provided ways to combine them~ the provided combinator functions to compose them

maybe?

Comment (ericmj) [reply _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: ~Generators implement the `Enumerable` protocol and they are infinite streams of terms~

Generators are infinite streams of terms that implement the `Enumerable` protocol

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 7]
```diff
@@ -0,0 +1,174 @@
+---
+layout: post
+title: StreamData: Property-based testing comes to Elixir
+author: Andrea Leopardi
+category: Releases
+excerpt: StreamData is a data-generation and property-based testing library that will be included in Elixir v1.6
+---
+
+We are happy to officially announce that Elixir v1.6 will come equipped with two new features: sample data generation as well as tools for property-based testing. We will mostly talk about property-based testing in this post, as sample data generation is a feature we introduced so that it could be leveraged by property-based testing. These functionalities are available today to developers (albeit in beta form) through the [stream_data][] library.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: I think we should decouple the announcement of StreamData with its inclusion in v1.6 for some reasons:

1. There is still a chance it won't be on v1.6 but rather on v1.7 (unlikely but possible)
2. Announcing that its inclusion is decided may actually demotivate people from giving feedback because they feel all decisions are taken

Given we want to encourage experimentation, we should reflect that there is space for that.

Comment (josevalim) [reply _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: > as sample data generation is a feature we introduced so that it could be leveraged by property-based testing

I think you are under selling sample data generation. :) It is OK to not cover sample data generation but it was implemented to be its own thing by design.

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+In this blog post, we'll cover what property-based testing is and how it can benefit your programs, why we want it in Elixir, and what we are exploring for the future of it. If you want to use the features discussed below or you want to read more formal documentation, head over to [stream_data][].
+
+## Sample data generation
+
+The core of the [stream_data][] library is `StreamData`: this module (which will be renamed to `Stream.Data` once the library is merged into Elixir) provides all the functionalities related to generating sample data of many kinds. It includes both data generators for data types (like integers or booleans) as well as tools to combine other generators (such as `one_of(list_of_generators)`).
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Let's cover the renaming of the modules in a separate section that talks about inclusion in v1.6 and what will change by then. This aligns well with my comment above.

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+`StreamData` contains some functions to modify generators. For example, you can build a generator of positive integers on top of `StreamData.integer()` and `StreamData.map/2`:
+
+```elixir
+generator = StreamData.map(StreamData.integer(), &abs/1)
+Enum.take(generator, 5)
+#=> [0, 1, 3, 3, 2]
+```
+
+`StreamData.map/2` is encouraged over `Stream.map/2` because generators only return "simple" terms when enumerated. When used in property-based testing, `StreamData` generators return wrapped values that contain the simple terms but also contain ways to *shrink* those terms, which is something property-based testing uses as we'll see later on.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: This reads a bit confusing for me. For the user perspective, they both "return" integers, right? Maybe it is best to drop the "simple" distinction and say that `StreamData` generators can shrink?

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+We decided to separate data-generation from property-based testing because it's something that developers can take advantage of in situations outside of property-based testing. For example, data streams can be used to seed a database or to have randomly generated data available during regular tests.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Beautiful!

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+```elixir
+test "length/1 calculates the length of a list" do
+  assert length([]) == 0
+  assert length([:one]) == 1
+  assert length([1, 2, 3]) == 3
+end
+```
+
+This test is written using an *example-based approach*. We are writing both the input to the piece of software we are testing as well as the expected output, and the testing tool is verifying that running the software on the given input results in the expected output. This style of testing is common and useful because it lets you get up and running easily and also lets you test known corner cases in an explicit way. However, this also means that it's hard to test many cases this way and even harder to uncover *unknown* corner cases that may reveal bugs in your code.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: > However, this also means that it's hard to test many cases this way and even harder to uncover *unknown* corner cases that may reveal bugs in your code.

There are too many "this". We can go with "this also means that it's hard to test many cases" OR "it's hard to test many cases this way".

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+```elixir
+property "length/1 is always >= 0" do
+  check all list <- list_of(term()) do
+    assert length(list) >= 0
+  end
+end
+```
+
+With property-based testing, you specify a set of valid inputs (lists in the example above) for your code and verify that your code holds some property for values taken at random from the valid inputs. In the example above, the test takes many (usually around 100) values at random from the `list_of(term())` *generator* and verifies that the property of `length/1` always returning a non-negative integer holds. A generator is just a `StreamData` generator, as we discussed in the previous section.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: "always returning" -> "always returns"

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+   Failed with generated values (after 4 attempt(s)):
+
+       list <- list_of(positive_integer())
+       #=> [4]
+
+   Expected false or nil, got true
+   code: refute Enum.any?(list, &(rem(&1, 4) == 0))
+```
+
+This error shows the minimal generated value that triggers the failure, that is, `[4]`. The process of finding the minimal generated value that triggers a failure is called *shrinking*. All generators that come with `StreamData` generate values that "bundle" a way to shrink them so that property-based testing can use this to provide the shrinking functionality. Shrinking is a fundamental part of property-based testing as it takes out the *noise* of random-generated data to reduce the failing data to focused and easier-to-understand terms.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Awesome! 🎉

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+### Using property-based testing in stream_data
+
+As you saw in the previous section, the core of property-based testing in `StreamData` is the `check all` macro. In this macro, you list a bunch of generators and filters (very similarly to how you would in `for` comprehensions) and then pass a body where you can verify that a property holds for the generated data.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Expressions like "As you saw in the previous section" are generally bad (at least I have been told so). If you just saw it, why are you repeating it? :)

Comment (josevalim) [reply _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: I would remove the `StreamData` module reference because the `check all` macro comes from elsewhere, so it looks a bit confusing (and not important in this context).

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+
+To make the `check all` macro available in your test, alongside importing all functions from `StreamData`, you can `use ExUnitProperties`:
+
+```elixir
+defmodule MyPropertyTest do
+  use ExUnit.Case, async: true
+  use ExUnitProperties
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Do we need to use or can we get away with an import?

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+              a > 0 and b > 0,
+              sum = a + b do
+      assert sum > a
+      assert sum > b            
+    end
+  end
+end
+```
+
+There's not much more to the mechanics of `StreamData`. Most of the work you will have to do revolves around finding good properties to test for your code and writing good generators for the data over which you want to test. Head over to [stream_data][]'s documentation for detailed documentation.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Same here in regards to `StreamData`. If you want to reference it, you should use `stream_data` and not the module but we already do that at the end of the paragraph.

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+              b <- integer(),
+              a > 0 and b > 0,
+              sum = a + b do
+      assert sum > a
+      assert sum > b            
+    end
+  end
+end
+```
+
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Maybe we can add that by using the property macro we can do things like `mix test --only property` to run only the properties in the code base?

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+A young but awesome book about property-based testing written by Fred Hebert is also available at [propertesting.com][]. This book is a *proper* (pun intended) guide to property-based testing and uses an Erlang library called [PropEr][]. However, the concepts and techniques perfectly apply to Elixir and `StreamData` as well.
+
+## Why include property-based testing in Elixir (and rewriting from scratch)
+
+The community has expressed some concern regarding two main things: why do we want to include a property-based testing tool in Elixir's standard library? And why write such a tool from scratch instead of using one of the existing Erlang or Elixir solutions?
+
+The answer to the first question is that we believe that providing such a tool in the standard library will encourage developers to use property-based testing and ultimately improve their software and the way they write it. At the same time, we want to be able to use property-based testing to test the Elixir codebase itself (which already turned out great [in the past][quickcheck-pr]). To do this, we can't rely on an external library, so we need to have a solution built into the standard library.
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Devil's advocate: why we can't rely on an external library? Theoretically, we could do it. :)

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown (hunk 15 above)
Comment (whatyouhide) [reply _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Well we have the `use` in the library so let's go with that for now. We can discuss if we want `import` both `ExUnitProperties` and `StreamData` as the "API" for this in stream_data though if you want to. Open an issue in case you want to :)

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 19]
```diff
@@ -0,0 +1,184 @@
+---
+layout: post
+title: StreamData: Property-based testing comes to Elixir
+author: Andrea Leopardi
+category: Releases
+excerpt: StreamData is a data-generation and property-based testing library that will be included in Elixir v1.6
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: "Included in Elixir v1.6" leftover.

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown (hunk 15 above)
Comment (josevalim) [reply _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: Ok, now i see why we need to use it. :+1:

File: _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+>
+> We understand rolling our own implementation has its downsides, especially since it lacks maturity compared to alternatives, but we balance it by actively seeking input from knowledgeable folks and by listening to the feedback that comes from the community, which we are very thankful for.
+
+## Roadmap
+
+stream_data and the functionalities it includes are scheduled to be included in one of the next two Elixir releases, likely 1.6 but possibly 1.7. The plan is to rename the `StreamData` module to `Stream.Data` and the `ExUnitProperties` module to `ExUnit.Properties`, so that the separation between data generation and property-based testing can be even more defined. Right now, all development is happening in the [stream_data][] repository, where we are discussing features and giving users a chance to try out the functionalities early on. We'd love for anyone to get involved in trying stream_data and we'd love feedback!
```
Comment (josevalim) [inline _posts/2017-10-21-stream-data-property-testing-comes-to-elixir.markdown]: "so that the separation between data generation and property-based testing can be even more defined" => to be fair, the reason we are renaming them is because the new names actually fit Elixir. So you could say that or you could say something like:

> Note we have also used the names `StreamData` and `ExUnitProperties` to avoid conflicts when those modules are eventually merged into Elixir. When merged, they will be renamed to the proper `Stream.Data` and `ExUnit.Properties` modules.
